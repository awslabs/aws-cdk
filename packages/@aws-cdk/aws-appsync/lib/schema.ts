import { IBucket } from '@aws-cdk/aws-s3';
import { CfnGraphQLSchema } from './appsync.generated';
import { GraphqlApi } from './graphqlapi';
import { AssetSchema, CodeSchema, S3Schema } from './private/schema-implementation';
import { IIntermediateType } from './schema-base';
import { ResolvableField } from './schema-field';
import { ObjectType } from './schema-intermediate';

/**
 * The Schema for a GraphQL Api
 *
 * If no options are configured, schema will be generated
 * code-first.
 */
export abstract class Schema {
  /**
   * Generate a Schema from file
   *
   * @returns `Schema` with immutable schema defintion
   * @param filePath the file path of the schema file
   */
  public static fromAsset(filePath: string): Schema {
    return new AssetSchema(filePath);
  }

  /**
   * Point GraphQL Schema to a S3 Location
   *
   * @returns `Schema` with immutable schema defintion
   * @param bucket the s3 bucket that holds the GraphQL Schema
   * @param key the object key for the schema
   */
  public static fromBucket(bucket: IBucket, key: string): Schema {
    return new S3Schema(bucket, key);
  }

  /**
   * Generate a Schema that can be edited through a code-first
   * approach
   *
   * @returns `Schema` with mutable schema defintion
   */
  public static fromCode(): Schema {
    return new CodeSchema();
  }

  /**
   * The definition for this schema
   */
  public readonly abstract definition: string;

  /**
   * Called when the GraphQL Api is initialized to allow this object to bind
   * to the stack.
   */
  public abstract bind(api: GraphqlApi): CfnGraphQLSchema;

  /**
   * Escape hatch to add to Schema as desired. Will always result
   * in a newline.
   */
  public abstract addToSchema(addition: string, delimiter?: string): void;
  /**
   * Add a query field to the schema's Query. CDK will create an
   * Object Type called 'Query'. For example,
   *
   * type Query {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addQuery(fieldName: string, field: ResolvableField): ObjectType;

  /**
   * Add a mutation field to the schema's Mutation. CDK will create an
   * Object Type called 'Mutation'. For example,
   *
   * type Mutation {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addMutation(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add a subscription field to the schema's Subscription. CDK will create an
   * Object Type called 'Subscription'. For example,
   *
   * type Subscription {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addSubscription(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add type to the schema
   */
  public abstract addType(_type: IIntermediateType): IIntermediateType;
}