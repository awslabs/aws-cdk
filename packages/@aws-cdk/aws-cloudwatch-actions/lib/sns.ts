import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import * as sns from '@aws-cdk/aws-sns';
import { Construct } from '@aws-cdk/core';

/**
 * Use an SNS topic as an alarm action
 */
export class SnsAction implements cloudwatch.IAlarmAction {
  constructor(private readonly topic: sns.ITopic) {
  }

  /**
   * Returns an alarm action configuration to use an SNS topic as an alarm action
   */
  public bind(_scope: Construct, _alarm: cloudwatch.IAlarm): cloudwatch.AlarmActionConfig {
    if (this.topic.masterKey) {
      this.topic.masterKey.grantEncryptDecrypt(new ServicePrincipal('cloudwatch.amazonaws.com'));
    }
    return { alarmActionArn: this.topic.topicArn };
  }
}
