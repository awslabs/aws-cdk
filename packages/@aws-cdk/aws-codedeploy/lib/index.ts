export * from './rollback-config';
export * from './ecs';
export * from './lambda';
export * from './server';
export * from './trigger-configuration';

// AWS::CodeDeploy CloudFormation Resources:
export * from './codedeploy.generated';
