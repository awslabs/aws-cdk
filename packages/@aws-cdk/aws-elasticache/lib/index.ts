// AWS::ElastiCache CloudFormation Resources:
export * from './elasticache.generated';
export * from './parameter-group';
