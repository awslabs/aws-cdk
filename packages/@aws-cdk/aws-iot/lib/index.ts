export * from './topic-rule';
export * from './topic-rule-action';
// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
