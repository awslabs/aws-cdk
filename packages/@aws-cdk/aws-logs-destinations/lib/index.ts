export * from './lambda';
export * from './kinesis';
export * from './kinesisfirehose';
