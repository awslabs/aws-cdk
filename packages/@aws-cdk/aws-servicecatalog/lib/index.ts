export * from './common';
export * from './constraints';
export * from './cloudformation-template';
export * from './portfolio';
export * from './product';
export * from './tagoptions';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
