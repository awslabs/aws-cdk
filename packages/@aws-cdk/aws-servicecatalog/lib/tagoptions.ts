import * as cdk from '@aws-cdk/core';
import { CfnTagOption } from '.';
import { hashValues } from './private/util';
import { InputValidator } from './private/validation';

/**
 * Defines a Tag Option, which are similar to tags
 * but have multiple values per key.
 */
export class TagOptions {
  /**
  * List of CfnTagOption
  */
  public readonly cfnTagOptions: CfnTagOption[];

  constructor(stack : cdk.Stack, tagOptionsMap: {[key: string]: string[]}) {
    this.cfnTagOptions = [];
    Object.keys(tagOptionsMap).forEach(key => {
      InputValidator.validateLength('TagOptions', 'TagOption key', 1, 128, key);
      tagOptionsMap[key].forEach((value: string) => {
        InputValidator.validateLength('TagOptions', 'TagOption value', 1, 256, value);
        const tagOptionKey = hashValues(key, value, stack.node.addr);
        const constructId = `TagOption${tagOptionKey}`;
        if (!stack.node.tryFindChild(constructId)) {
          const cfnTagOption = new CfnTagOption(stack, constructId, {
            key: key,
            value: value,
            active: true,
          });
          this.cfnTagOptions.push(cfnTagOption);
        }
      });
    });
  }
}
