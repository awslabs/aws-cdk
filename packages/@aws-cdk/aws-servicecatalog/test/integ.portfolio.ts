import * as iam from '@aws-cdk/aws-iam';
import * as sns from '@aws-cdk/aws-sns';
import { App, Stack } from '@aws-cdk/core';
import * as servicecatalog from '../lib';

const app = new App();
const stack = new Stack(app, 'integ-servicecatalog-portfolio');

const role = new iam.Role(stack, 'TestRole', {
  assumedBy: new iam.AccountRootPrincipal(),
});

const group = new iam.Group(stack, 'TestGroup');

const portfolio = new servicecatalog.Portfolio(stack, 'TestPortfolio', {
  displayName: 'TestPortfolio',
  providerName: 'TestProvider',
  description: 'This is our Service Catalog Portfolio',
  messageLanguage: servicecatalog.MessageLanguage.EN,
});

portfolio.giveAccessToRole(role);
portfolio.giveAccessToGroup(group);

portfolio.shareWithAccount('123456789012');

const product = new servicecatalog.CloudFormationProduct(stack, 'TestProduct', {
  productName: 'testProduct',
  owner: 'testOwner',
  productVersions: [
    {
      validateTemplate: false,
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromUrl(
        'https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template'),
    },
  ],
});

portfolio.addProduct(product);

portfolio.constrainTagUpdates(product);

const topics = [
  new sns.Topic(stack, 'Topic1'),
  new sns.Topic(stack, 'Topic2'),
  new sns.Topic(stack, 'Topic3'),
];
const specialTopic = new sns.Topic(stack, 'specialTopic');

portfolio.addEventNotifications(product, topics);
portfolio.addEventNotifications(product, [specialTopic], {
  description: 'special topic description',
  messageLanguage: servicecatalog.MessageLanguage.EN,
});

app.synth();
