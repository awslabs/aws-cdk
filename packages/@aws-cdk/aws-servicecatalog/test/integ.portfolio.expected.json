{
  "Resources": {
    "TestRole6C9272DF": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "TestGroupAF88660E": {
      "Type": "AWS::IAM::Group"
    },
    "TestPortfolio4AC794EB": {
      "Type": "AWS::ServiceCatalog::Portfolio",
      "Properties": {
        "DisplayName": "TestPortfolio",
        "ProviderName": "TestProvider",
        "AcceptLanguage": "en",
        "Description": "This is our Service Catalog Portfolio"
      }
    },
    "TestPortfolioPortolioPrincipalAssociation20e1afa20ac27E1A060D": {
      "Type": "AWS::ServiceCatalog::PortfolioPrincipalAssociation",
      "Properties": {
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "PrincipalARN": {
          "Fn::GetAtt": [
            "TestRole6C9272DF",
            "Arn"
          ]
        },
        "PrincipalType": "IAM"
      }
    },
    "TestPortfolioPortolioPrincipalAssociation44a1ca1c23384D6E460B": {
      "Type": "AWS::ServiceCatalog::PortfolioPrincipalAssociation",
      "Properties": {
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "PrincipalARN": {
          "Fn::GetAtt": [
            "TestGroupAF88660E",
            "Arn"
          ]
        },
        "PrincipalType": "IAM"
      }
    },
    "TestPortfolioPortfolioSharebf5b82f042508F035880": {
      "Type": "AWS::ServiceCatalog::PortfolioShare",
      "Properties": {
        "AccountId": "123456789012",
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        }
      }
    },
    "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7": {
      "Type": "AWS::ServiceCatalog::PortfolioProductAssociation",
      "Properties": {
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        }
      }
    },
    "TestPortfolioResourceUpdateConstrainta0185761d231AB0EAAE0": {
      "Type": "AWS::ServiceCatalog::ResourceUpdateConstraint",
      "Properties": {
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        },
        "TagUpdateOnProvisionedProduct": "ALLOWED"
      },
      "DependsOn": [
        "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7"
      ]
    },
    "TestPortfolioLaunchNotificationConstraint177f7a17cc920269D7A5": {
      "Type": "AWS::ServiceCatalog::LaunchNotificationConstraint",
      "Properties": {
        "NotificationArns": [
          {
            "Ref": "Topic198E71B3E"
          }
        ],
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        }
      },
      "DependsOn": [
        "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7"
      ]
    },
    "TestPortfolioLaunchNotificationConstraintbb8276cabdbf6C57454B": {
      "Type": "AWS::ServiceCatalog::LaunchNotificationConstraint",
      "Properties": {
        "NotificationArns": [
          {
            "Ref": "Topic269377B75"
          }
        ],
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        }
      },
      "DependsOn": [
        "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7"
      ]
    },
    "TestPortfolioLaunchNotificationConstraint33147fa9fb6e1F5CF9BA": {
      "Type": "AWS::ServiceCatalog::LaunchNotificationConstraint",
      "Properties": {
        "NotificationArns": [
          {
            "Ref": "Topic3DEAE47A7"
          }
        ],
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        }
      },
      "DependsOn": [
        "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7"
      ]
    },
    "TestPortfolioLaunchNotificationConstraint637de3424f24E7D044EA": {
      "Type": "AWS::ServiceCatalog::LaunchNotificationConstraint",
      "Properties": {
        "NotificationArns": [
          {
            "Ref": "specialTopic7664DE4C"
          }
        ],
        "PortfolioId": {
          "Ref": "TestPortfolio4AC794EB"
        },
        "ProductId": {
          "Ref": "TestProduct7606930B"
        },
        "AcceptLanguage": "en",
        "Description": "special topic description"
      },
      "DependsOn": [
        "TestPortfolioPortfolioProductAssociationa0185761d231B0D998A7"
      ]
    },
    "TestProduct7606930B": {
      "Type": "AWS::ServiceCatalog::CloudFormationProduct",
      "Properties": {
        "Name": "testProduct",
        "Owner": "testOwner",
        "ProvisioningArtifactParameters": [
          {
            "DisableTemplateValidation": true,
            "Info": {
              "LoadTemplateFromURL": "https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template"
            }
          }
        ]
      }
    },
    "Topic198E71B3E": {
      "Type": "AWS::SNS::Topic"
    },
    "Topic269377B75": {
      "Type": "AWS::SNS::Topic"
    },
    "Topic3DEAE47A7": {
      "Type": "AWS::SNS::Topic"
    },
    "specialTopic7664DE4C": {
      "Type": "AWS::SNS::Topic"
    }
  }
}